zipline/pipeline/hooks/progress.py

```diff
@@ -0,0 +1,494 @@
+    current_work : [zipline.pipeline.term.Term]
+        List of terms currently being loaded or computed.
+    """
+
+    def __init__(self, nterms, start_date, end_date):
+        self._start_date = start_date
+        self._end_date = end_date
+
+        # +1 to be inclusive of end_date.
+        total_days = (end_date - start_date).days + 1
+        self._max_progress = total_days * nterms
+
+        self._progress = 0
+        self._days_completed = 0
+
+        self._state = 'init'
+
+        self._current_chunk_size = None
+        self._current_chunk_bounds = None
+        self._current_work = None
+
+        self._start_time = time.time()
+        self._end_time = None
+
+    # These properties form the public interface for Publishers.
+    @property
+    def state(self):
+        return self._state
+
+    @property
+    def percent_complete(self):
+        return 100.0 * float(self._progress) / self._max_progress
+
+    @property
+    def execution_time(self):
+        if self._end_time is None:
+            raise ValueError(
+                "Can't get execution_time until execution is complete."
+            )
+        return self._end_time - self._start_time
+
+    @property
+    def execution_bounds(self):
+        return (self._start_date, self._end_date)
+
+    @property
+    def current_chunk_bounds(self):
+        return self._current_chunk_bounds
+
+    @property
+    def current_work(self):
+        return self._current_work
+
+    # These methods form the interface for ProgressHooks.
+    def start_chunk(self, start_date, end_date):
+        days_since_start = (end_date - self._start_date).days + 1
+        self._current_chunk_size = days_since_start - self._days_completed
+        self._current_chunk_bounds = (start_date, end_date)
+
+    def finish_chunk(self, start_date, end_date):
+        self._days_completed += self._current_chunk_size
+
+    def start_load_terms(self, terms):
+        self._state = 'loading'
+        self._current_work = terms
+
+    def finish_load_terms(self, terms):
+        self._increment_progress(nterms=len(terms))
+
+    def start_compute_term(self, term):
+        self._state = 'computing'
+        self._current_work = [term]
+
+    def finish_compute_term(self, term):
+        self._increment_progress(nterms=1)
+
+    def finish(self, success):
+        self._end_time = time.time()
+        if success:
+            self._state = 'success'
+        else:
+            self._state = 'error'
+
+    def _increment_progress(self, nterms):
+        self._progress += nterms * self._current_chunk_size
+
+
+try:
+    import ipywidgets
+    HAVE_WIDGETS = True
+except ImportError:
+    HAVE_WIDGETS = False
+
+try:
+    from IPython.display import display, HTML as IPython_HTML
+    HAVE_IPYTHON = True
+except ImportError:
+    HAVE_IPYTHON = False
+
+
+# XXX: This class is currently untested, because we don't require ipywidgets as
+#      a test dependency. Be careful if you make changes to this.
```

Comment:
Do you want me to look hard at this? I feel like if it looks good then it is "correct".

Reply:
I think this is good for now. As we add more functionality to this, I think I'll want to add a proper test suite for this, but for now I don't want to deal with adding ipywidgets to CI.